### Utilizando a base de dados archive.ics.uci.edu/ml/datasets/Car+Evaluation:

In [39]:
import pandas as pd
import numpy as np

data = pd.read_csv('car.csv')
data.head()

buying  maint doors persons lug_boot safety  class
0  vhigh  vhigh     2       2    small    low  unacc
1  vhigh  vhigh     2       2    small    med  unacc
2  vhigh  vhigh     2       2    small   high  unacc
3  vhigh  vhigh     2       2      med    low  unacc
4  vhigh  vhigh     2       2      med    med  unacc

### 1. (10 pontos) Defina o classificador cego com o menor erro de classificação possível. 

##### calcular a probabilidade a priori de cada classe

In [9]:
tot  = data['class'].size
tot

1728

In [10]:
cont = data['class'].value_counts()
cont

unacc    1210
acc       384
good       69
vgood      65
Name: class, dtype: int64

In [16]:
sum = 0
for i in cont:
    prob = i/tot
    print(prob)
    sum = sum + prob
    
sum

0.7002314814814815
0.2222222222222222
0.03993055555555555
0.03761574074074074


1.0

##### Erro maximo para cada classe

In [18]:
#unacc
1 - (cont[0]/tot)

0.2997685185185185

In [21]:
#acc
1 - (cont[1]/tot)

0.7777777777777778

In [22]:
#good
1 - (cont[2]/tot)

0.9600694444444444

In [25]:
#vgood
1 - (cont[3]/tot)

0.9623842592592593

<font color="yellow">Resposta: O classificador cego com o menor erro de classificação possível é unacc (Erro máximo = 0.2997685185185185)</font>

### 2. (18 pontos) Calcule as probablidades a priori para cada variável (para cada possível valor de cada variável).

In [73]:
for x in data.drop(['class'], axis=1):
    unique_values = np.unique(data[x].values)
    print("\n" + x + "\n")
    for y in unique_values:
        print(y, str(data[x].value_counts()[y]) + "/" + str(data[x].size))


buying

high 432/1728
low 432/1728
med 432/1728
vhigh 432/1728

maint

high 432/1728
low 432/1728
med 432/1728
vhigh 432/1728

doors

2 432/1728
3 432/1728
4 432/1728
5more 432/1728

persons

2 576/1728
4 576/1728
more 576/1728

lug_boot

big 576/1728
med 576/1728
small 576/1728

safety

high 576/1728
low 576/1728
med 576/1728


### 3. (18 pontos) Calcule a probabilidade de cada valor de cada variável dado a classe (posteriori).

In [70]:
class_values = np.unique(data['class'].values)

for x in data.drop(["class"], axis=1):
    unique_values = np.unique(data[x].values)
    print("\n" + x + "\n")
    for y in unique_values:
        for i in class_values:
            try:
                print(y + " | " + i, str(data.groupby(x)['class'].value_counts()[y][i]) + "/" + str(data['class'].value_counts()[i]))
            except:
                next




buying

high | acc 108/384
high | unacc 324/1210
low | acc 89/384
low | good 46/69
low | unacc 258/1210
low | vgood 39/65
med | acc 115/384
med | good 23/69
med | unacc 268/1210
med | vgood 26/65
vhigh | acc 72/384
vhigh | unacc 360/1210

maint

high | acc 105/384
high | unacc 314/1210
high | vgood 13/65
low | acc 92/384
low | good 46/69
low | unacc 268/1210
low | vgood 26/65
med | acc 115/384
med | good 23/69
med | unacc 268/1210
med | vgood 26/65
vhigh | acc 72/384
vhigh | unacc 360/1210

doors

2 | acc 81/384
2 | good 15/69
2 | unacc 326/1210
2 | vgood 10/65
3 | acc 99/384
3 | good 18/69
3 | unacc 300/1210
3 | vgood 15/65
4 | acc 102/384
4 | good 18/69
4 | unacc 292/1210
4 | vgood 20/65
5more | acc 102/384
5more | good 18/69
5more | unacc 292/1210
5more | vgood 20/65

persons

2 | unacc 576/1210
4 | acc 198/384
4 | good 36/69
4 | unacc 312/1210
4 | vgood 30/65
more | acc 186/384
more | good 33/69
more | unacc 322/1210
more | vgood 35/65

lug_boot

big | acc 144/384
big | good 24/69

### 4. (18 pontos) Implemente o classificador Naive Bayes.

### 5. (18 pontos) Utilizando sua implementação e metade dos exemplos de cada classe como conjunto de treino e o resante como conjunto de teste, avalie o classificador Naive Bayes.

### 6. (18 pontos) Faça o mesmo da questão anterior utilizando uma implementação de biblioteca com configuração semelhente ao classificador solicitado na questão anterior.